File: src/main/java/org/example/kalkulationsprogramm/service/GewerkService.java
package org.example.kalkulationsprogramm.service;

import lombok.RequiredArgsConstructor;
import org.example.kalkulationsprogramm.domain.Gewerk;
import org.example.kalkulationsprogramm.dto.GewerkDto;
import org.example.kalkulationsprogramm.repository.GewerkRepository;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

@Service
@RequiredArgsConstructor
public class GewerkService {

    private final GewerkRepository repository;

    @Transactional(readOnly = true)
    public List<GewerkDto> findAll() {
        return repository.findAllByOrderByNameAsc().stream().map(GewerkService::toDto).toList();
    }

    @Transactional(readOnly = true)
    public List<GewerkDto> findAktiv() {
        return repository.findByAktivTrueOrderByNameAsc().stream().map(GewerkService::toDto).toList();
    }

    @Transactional
    public GewerkDto save(GewerkDto dto) {
        Gewerk entity = dto.getId() != null
                ? repository.findById(dto.getId()).orElseThrow(() -> new IllegalArgumentException("Gewerk nicht gefunden: " + dto.getId()))
                : new Gewerk();
        if (dto.getName() == null || dto.getName().isBlank()) {
            throw new IllegalArgumentException("Name darf nicht leer sein.");
        }
        if (dto.getBgName() == null || dto.getBgName().isBlank()) {
            throw new IllegalArgumentException("BG-Name darf nicht leer sein.");
        }
        if (dto.getBgSatzProzent() == null) {
            throw new IllegalArgumentException("BG-Satz (Prozent) ist Pflicht.");
        }
        entity.setName(dto.getName().trim());
        entity.setBgName(dto.getBgName().trim());
        entity.setBgSatzProzent(dto.getBgSatzProzent());
        entity.setAktiv(dto.getAktiv() == null ? Boolean.TRUE : dto.getAktiv());
        entity.setBemerkung(dto.getBemerkung());
        return toDto(repository.save(entity));
    }

    @Transactional
    public void delete(Long id) {
        repository.deleteById(id);
    }

    public static GewerkDto toDto(Gewerk e) {
        GewerkDto dto = new GewerkDto();
        dto.setId(e.getId());
        dto.setName(e.getName());
        dto.setBgName(e.getBgName());
        dto.setBgSatzProzent(e.getBgSatzProzent());
        dto.setAktiv(e.getAktiv());
        dto.setBemerkung(e.getBemerkung());
        return dto;
    }
}

File: src/main/kotlin/org/example/kalkulationsprogramm/service/GewerkService.kt
package org.example.kalkulationsprogramm.service

import org.example.kalkulationsprogramm.domain.Gewerk
import org.example.kalkulationsprogramm.dto.GewerkDto
import org.example.kalkulationsprogramm.repository.GewerkRepository
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
class GewerkService(
    private val repository: GewerkRepository
) {
    @Transactional(readOnly = true)
    fun findAll(): List<GewerkDto> = repository.findAllByOrderByNameAsc().map(::toDto)

    @Transactional(readOnly = true)
    fun findAktiv(): List<GewerkDto> = repository.findByAktivTrueOrderByNameAsc().map(::toDto)

    @Transactional
    fun save(dto: GewerkDto): GewerkDto {
        val entity = dto.id?.let {
            repository.findById(it).orElseThrow { IllegalArgumentException("Gewerk nicht gefunden: ${dto.id}") }
        } ?: Gewerk()
        val name = dto.name?.trim()
        val bgName = dto.bgName?.trim()
        require(!name.isNullOrBlank()) { "Name darf nicht leer sein." }
        require(!bgName.isNullOrBlank()) { "BG-Name darf nicht leer sein." }
        require(dto.bgSatzProzent != null) { "BG-Satz (Prozent) ist Pflicht." }
        entity.name = name
        entity.bgName = bgName
        entity.bgSatzProzent = dto.bgSatzProzent
        entity.aktiv = dto.aktiv ?: true
        entity.bemerkung = dto.bemerkung
        return toDto(repository.save(entity))
    }

    @Transactional
    fun delete(id: Long) {
        repository.deleteById(id)
    }

    companion object {
        @JvmStatic
        fun toDto(e: Gewerk): GewerkDto {
            val dto = GewerkDto()
            dto.id = e.id
            dto.name = e.name
            dto.bgName = e.bgName
            dto.bgSatzProzent = e.bgSatzProzent
            dto.aktiv = e.aktiv
            dto.bemerkung = e.bemerkung
            return dto
        }
    }
}
